### PR TITLE
feat(vault-index): FTS5 hybrid reranking with AND queries and 5-signal scorer

### PR DESCRIPTION
**Pull request by abhattacherjee**
## Summary
- Switch FTS5 search from OR-mode to AND-mode queries with automatic OR fallback when AND returns zero results
- Add BM25 column weighting (title 10x, tags 5x, body 1x) for relevance-based ranking
- Implement 5-signal Python reranker: term proximity (0.35), BM25 (0.25), note type (0.15), recency (0.15), term density (0.10)
- Add `body` column to `notes` table with auto-migration detection via `PRAGMA table_info`

## Context
Multi-term vault searches like "sentry feasibility" returned hundreds of irrelevant results because FTS5 used OR-joined queries. This two-stage pipeline (FTS5 broad retrieval → Python reranking) ensures relevant results surface first while maintaining O(log n) performance.

## Review Rounds
- **Subagent reviews:** 2 spec compliance + 2 code quality → empty-phrase filter fix
- **Copilot Round 1:** 5 comments → 3 fixed (density word boundaries, migration check, docstring), 2 acknowledged
- **3-agent parallel review:** code-reviewer + silent-failure-hunter + test-analyzer → Layer 3 OR-mode regression fix, proximity dedup bug fix, OR fallback logging, DB delete error logging, 14 edge case tests

## Test Plan
- [x] 35 new tests across 7 test classes (375 total, 90%+ coverage)
- [x] Schema migration: old DB without body column auto-rebuilt
- [x] AND-mode: multi-word, phrase preservation, hyphen handling, empty query, unmatched quotes
- [x] BM25 title boost: title match ranks above body-only match
- [x] OR fallback: partial matches returned when AND finds nothing
- [x] Reranker signals: proximity, type boost, recency, density, limit enforcement
- [x] Edge cases: None body/date, empty query terms, duplicate terms, body stripping
- [x] Integration: "sentry feasibility" returns insight first, no noise notes
- [x] Regression: `query_related_notes` Layer 3 OR-mode with partial keyword matches

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: hooks/vault_index.py [hunk 1]
```diff
@@ -301,6 +310,17 @@ def ensure_index(vault_path: str, folders: list[str], db_path: str | None = None
         conn = _connect(db_path)
         try:
             _init_schema(conn)
+            if _needs_body_migration(conn):
+                print(f"[vault-index] Missing body column; rebuilding {db_path}",
+                      file=sys.stderr)
+                conn.close()
+                for suffix in ("", "-wal", "-shm"):
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/vault_index.py:317]: Fixed in d505e65 — added post-rebuild assertion verifying body column exists after migration.

File: hooks/vault_index.py [hunk 2]
```diff
@@ -190,8 +198,8 @@ def _upsert_note(conn: sqlite3.Connection, rel_path: str, parsed: dict, mtime: f
 
     conn.execute(
         "INSERT INTO notes (path, type, date, project, title, source_session, "
-        "source_note, tags, status, mtime, size) "
-        "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
+        "source_note, tags, status, mtime, size, body) "
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/vault_index.py:201]: Known design decision for contentless FTS5. Orphaned entries are invisible (filtered by JOIN) and cleaned by rebuild_index(). See spec section 'Why not content-backed FTS5?' — the orphaned-entry strategy is simpler and battle-tested.

File: hooks/vault_index.py [hunk 3]
```diff
@@ -396,32 +416,179 @@ def index_note(db_path: str, note_path: str) -> bool:
...
+    if len(positions) < 2:
+        return 0.0
+
+    min_dist = float("inf")
+    terms_with_pos = list(positions.keys())
+    for i in range(len(terms_with_pos)):
+        for j in range(i + 1, len(terms_with_pos)):
+            for p1 in positions[terms_with_pos[i]]:
+                for p2 in positions[terms_with_pos[j]]:
+                    dist = abs(p1 - p2)
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/vault_index.py:507]: Acknowledged. With 50 candidates and typical notes under 5KB, the nested loop is ~5ms total. Two-pointer merge would save sub-milliseconds. Will revisit if vault scales past 10K notes.

File: hooks/vault_index.py [hunk 4]
```diff
@@ -396,32 +416,179 @@ def index_note(db_path: str, note_path: str) -> bool:
...
+        try:
+            note_date = date.fromisoformat(r.get("date", "") or "")
+            days_old = max((today - note_date).days, 0)
+        except (ValueError, TypeError):
+            days_old = 365
+        recency = 0.5 ** (days_old / 30)
+
+        if query_terms:
+            matched = sum(1 for t in query_terms if t in full_text)
+            density = matched / len(query_terms)
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/vault_index.py:569]: Good catch — fixed in d505e65. Density now uses regex word boundaries: re.search(r'\\b' + re.escape(t) + r'\\b', full_text).

File: hooks/vault_index.py [hunk 5]
```diff
@@ -396,32 +416,179 @@ def index_note(db_path: str, note_path: str) -> bool:
...
     note_type: str | None = None,
-    limit: int = 20,
+    limit: int = 15,
 ) -> list[dict]:
     """Full-text search over the vault index.
 
-    Returns list of dicts with note metadata + rank.
+    Returns list of dicts with note metadata + rank (body excluded).
+    Uses BM25 column weighting (title=10x, tags=5x) and falls back to
+    OR-mode if the AND query returns no results.
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/vault_index.py]: Fixed in d505e65 — docstring now documents rerank_score, default limit=15, body stripping, BM25 weighting, and OR fallback.
